### PR TITLE
Ipynb benchmarking

### PR DESCRIPTION
**Pull request by aqitya**
Jupyter Notebook for Benchmarking Example.

### CODE REVIEW

File: examples/benchmarking.ipynb [hunk 1]
```diff
@@ -0,0 +1,180 @@
+{
+ "cells": [
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%matplotlib inline"
```
Comment (kjjarvis) [inline examples/benchmarking.ipynb:9]: Is this supposed to show up at the top? (the answer might be yes, I'm not very familiar enough with jupyter notebooks)

File: examples/benchmarking.ipynb [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+   "source": [
+    "# Welcome to ProgPy's Benchmarking Example"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The goal of this notebook is to demonstrate the Benchmarking feature offered for Prognostic Models. Specifically, we will demonstrate via a simple example how to benchmark the computational efficiency of models!"
```
Comment (kjjarvis) [inline examples/benchmarking.ipynb:25]:
```suggestion
    "The goal of this notebook is to demonstrate the Benchmarking feature offered for Prognostic Models. Specifically, we will demonstrate how to benchmark the computational efficiency of models with a simple example."
```

File: examples/benchmarking.ipynb [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+   "source": [
+    "The first import is importing ProgPy's BatteryCircuit Model, and the second import will be used to benchmark our model!"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Next, let's initialize our Battery Circuit by creating model object."
```
Comment (kjjarvis) [inline examples/benchmarking.ipynb:59]:
```suggestion
    "Next, let's initialize our Battery Circuit by creating a model object."
```

File: examples/benchmarking.ipynb [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+    "# Step 1: Create a model object\n",
+    "batt = BatteryCircuit()"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Then, for our model, we would need to define a future loading function. More information on what a future loading function is and how to use it can be found here: https://nasa.github.io/progpy/prog_models_guide.html#future-loading"
```
Comment (kjjarvis) [inline examples/benchmarking.ipynb:77]:
```suggestion
    "Then, for our model, we will need to define a future loading function. More information on what a future loading function is and how to use it can be found here: https://nasa.github.io/progpy/prog_models_guide.html#future-loading"
```

File: examples/benchmarking.ipynb [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+    "    # Constant Loading\n",
+    "    return loading"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Since this is a simple example, we are going to have a constant loading!"
```
Comment (kjjarvis) [inline examples/benchmarking.ipynb:98]: Maybe move this to the block above the future_loading code?

File: examples/benchmarking.ipynb [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Finally, we are going to need to benchmark the simulation!\n",
```
Comment (kjjarvis) [inline examples/benchmarking.ipynb:106]:
```suggestion
    "Finally, we are ready to benchmark the simulation.\n",
```

File: examples/benchmarking.ipynb [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+    "\n",
+    "We can do this by using the `timeit()` function and pass in our `simulate_to()` or `simulate_to_threshold()` function for the `stmt` argument. For more information regarding the `timeit()` function, please read it's documentation located here: https://docs.python.org/3/library/timeit.html"
```
Comment (kjjarvis) [inline examples/benchmarking.ipynb:108]:
```suggestion
    "We can do this by using the `timeit()` function and pass in our `simulate_to()` or `simulate_to_threshold()` function for the `stmt` argument. For more information regarding the `timeit()` function, please read its documentation located here: https://docs.python.org/3/library/timeit.html"
```

Comment (kjjarvis) [reply examples/benchmarking.ipynb:108]: Would there be another way to write: "function for the 'stmt' argument" that would convey the appropriate information? I didn't know what this meant. It's clear once I looked up timeit() (and maybe is clear to more code-oriented people), but just a suggestion to maybe reword for accessibility to all users.

File: examples/benchmarking.ipynb [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+    "    batt.simulate_to(600, future_loading)\n",
+    "time = timeit(sim, number=500)"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "In this example, we would be benchmarking the simulation for the BatteryCircuit model up to 600 seconds. Furthermore, we will define our `number` argument to be 500 for sake of runtime."
```
Comment (kjjarvis) [inline examples/benchmarking.ipynb:128]:
```suggestion
    "In this example, we are benchmarking the simulation for the BatteryCircuit model up to 600 seconds. Furthermore, we define our `number` argument to be 500 for sake of runtime."
```

Comment (kjjarvis) [reply examples/benchmarking.ipynb:128]: Is there a way to reword "number argument" to make it more descriptive? Or perhaps we can add a sentence describing what exactly is set to 500

File: examples/benchmarking.ipynb (hunk 2 above)
Comment (kjjarvis) [reply examples/benchmarking.ipynb:25]: Now that we have combined prog_models and prog_algs, I think we need to make a clear distinction between this benchmarking and that in benchmarking_example. It might be useful to add a sentence about what 'benchmarking' is in this case. We could potentially even rename one of these to make the distinction clear. Let's chat about this in more detail.

File: examples/benchmarking.ipynb [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+    "# Print results\n",
+    "print('Simulation Time: {} ms/sim'.format(time))"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "In this example, we benchmarked the simulation of the BatteryCircuit model up to 600 seconds. By utilizing the `time` package, we are able to benchmark the simulation of the model and print out the results with ease!"
```
Comment (kjjarvis) [inline examples/benchmarking.ipynb:154]: These two sentences feel a tad repetitive. Maybe we could adjust one to have a description of what benchmarking means in this context? Open to your suggestions.

**Review by kjjarvis — CHANGES_REQUESTED** [review]
Nice start! This looks very good. Here are a few comments, and we'll also need to address the extra files (I think Chris said he'd help with this).

File: examples/benchmarking.ipynb (hunk 9 above)
Comment (aqitya) [reply examples/benchmarking.ipynb:154]: Yeah, I agree, how does this sound:

'In this example, we benchmarked the simulation of the BatteryCircuit model up to 600 seconds by utilizing the `time` package!'

Comment (kjjarvis) [reply examples/benchmarking.ipynb]: Sounds good, nice job.

**Review by kjjarvis — CHANGES_REQUESTED** [review]
@aqitya there are still some outstanding comments from the last review that need to be addressed. Please take a look, and let me know if you don't agree and we can brainstorm solutions together.

File: examples/benchmarking.ipynb (hunk 1 above)
Comment (teubert) [reply examples/benchmarking.ipynb:9]: I remember us talking about it at one point in the past, but I can't remember why it's here. Can you explain it for me?

File: examples/benchmarking.ipynb [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+   "source": [
+    "# Welcome to ProgPy's Benchmarking Example"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The goal of this notebook is to demonstrate the Benchmarking feature offered for Prognostic Models. Specifically, we will demonstrate how to benchmark the computational efficiency of models with a simple example."
```
Comment (teubert) [inline examples/benchmarking.ipynb:25]:
```suggestion
    "The goal of this notebook is to demonstrate benchmarking Prognostic Models. Specifically, we will demonstrate how to benchmark the computational efficiency of model simulation. This is typically what you want to look at when benchmarking models, since simulation is the foundation of state estimation and prediction."
```

It's not strictly a ProgPy feature, but demonstrating how to benchmark a model using other tools

File: examples/benchmarking.ipynb [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+    "First, we need to import the necessary modules."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from prog_models.models import BatteryCircuit\n",
```
Comment (teubert) [inline examples/benchmarking.ipynb:42]:
```suggestion
    "from progpy.models import BatteryCircuit\n",
```

File: examples/benchmarking.ipynb [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+    "from timeit import timeit"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The first import is importing ProgPy's BatteryCircuit Model, and the second import will be used to benchmark our model!"
```
Comment (teubert) [inline examples/benchmarking.ipynb:51]:
```suggestion
    "The first import is importing a model to benchmark. In this case, ProgPy's BatteryCircuit Model. The second import is of the timeit tool, which will be used to benchmark our model."
```

File: examples/benchmarking.ipynb [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Next, let's initialize our Battery Circuit by creating a model object."
```
Comment (teubert) [inline examples/benchmarking.ipynb:59]:
```suggestion
    "Next, let's create our Battery Circuit model."
```
Initialization is something else in ProgPy (.initialize()). Though correct, I want to avoid using that term to not confuse

**Review by teubert — CHANGES_REQUESTED** [review]
Looking good. A few suggestions

File: examples/benchmarking.ipynb (hunk 1 above)
Comment (aqitya) [reply examples/benchmarking.ipynb:9]: We mentioned how the automated notebooks had this at the top. But from what I can tell, this does not serve a purpose for us, so I'll remove this cell.

**Review by kjjarvis — APPROVED** [review]
This notebook is looking great, approved. One idea for the future - now that ProgPy is combined, we have two 'benchmarking' examples. Should we rename these files with something more specific for clarity?

**Review by teubert — APPROVED** [review]
Looks great. Good work!
